interfaces/prompting/requestrules/requestrules.go

```diff
@@ -0,0 +1,877 @@
+// conflicting pattern variants and permissions by pruning the offending
+//
+// This function is only required if database is left inconsistent (should not
+// occur) or when loading, in case the stored rules on disk were corrupted.
+//
+// By default, issues a notice for each rule which is modified or removed as a
+// result of a conflict with another rule. If notifyEveryRule is true, issues
+// a notice for every rule which was in the database prior to the beginning of
+// the function. In either case, at most one notice is issued for each rule.
+//
+// Discards the current rule tree, then iterates through the rules in rdb.rules
+// and re-populates the tree. If there are any conflicts between rules (that
+// is, rules share the same path pattern and one or more of the same
+// permissions), the conflicting permission is removed from the rule with the
+// earlier timestamp. When the function returns, the database should be fully
+// internally consistent and without conflicting rules.
+//
+// TODO: unexport (probably)
+func (rdb *RuleDB) RefreshTreeEnforceConsistency(notifyEveryRule bool) {
+	rdb.mutex.Lock()
+	defer rdb.mutex.Unlock()
+	needToSave := false
+	defer func() {
+		if needToSave {
+			rdb.save()
+		}
+	}()
+	modifiedUserRuleIDs := make(map[uint32]map[prompting.IDType]map[string]string)
+	defer func() {
+		for user, ruleIDs := range modifiedUserRuleIDs {
+			for ruleID, data := range ruleIDs {
+				rdb.notifyRule(user, ruleID, data)
+			}
+		}
+	}()
+	currTime := time.Now()
+	origRules := rdb.rules
+	rdb.ids = make(map[prompting.IDType]int)
+	rdb.rules = make([]*Rule, 0, len(rdb.rules))
+	rdb.perUser = make(map[uint32]*userDB)
+	for _, rule := range origRules {
+		_, exists := modifiedUserRuleIDs[rule.User]
+		if !exists {
+			modifiedUserRuleIDs[rule.User] = make(map[prompting.IDType]map[string]string)
+		}
+		if notifyEveryRule {
+			modifiedUserRuleIDs[rule.User][rule.ID] = nil
+		}
+		if err := rule.Lifespan.ValidateExpiration(rule.Expiration, currTime); err != nil || rule.Constraints.ValidateForInterface(rule.Interface) != nil {
+			// Invalid rule, discard it
+			needToSave = true
+			data := map[string]string{"removed": "invalid"}
+			if errors.Is(err, prompting.ErrExpirationInThePast) {
+				// Not actually invalid, just expired
+				data["removed"] = "expired"
+			}
+			modifiedUserRuleIDs[rule.User][rule.ID] = data
+			continue
+		}
+		for {
+			err, conflicts, conflictingPermission := rdb.addRuleToTree(rule)
+			if err == nil {
+				break
+			}
+			// Err must be ErrPathPatternConflict.
+			// Prioritize newer rules by pruning permission from old rule until
+			// no conflicts remain.
+			// XXX: this results in the permission being dropped for all other
+			// variants of the older rule.
+			// TODO: split older rule into two rules, preserving all except the
+			// directly conflicting variant/permission combination.
+			for _, conflict := range conflicts {
```

Comment:
do we need to bit in practice?

Reply:
We need basic conflict resolution (what's currently implemented) in case the rule DB on disk has conflicts in it, since we can't allow conflicts in the in-memory DB. We don't need the fancy inclusion-exclusion pattern splitting mentioned in the TODO, though technically doing something like that would lead to optimal coverage of the desired rules (assuming timestamps are correct), though it would mangle the existing rules to do so.

I suppose the easy solution would be that if there is a conflict, throw away the rule DB and start over. That would greatly simplify things, though it would make it more likely the rules DB is wiped.